File: Days.30/Python/3.Quick-Sort-Array.py
# Quick Sort using Haore's Parition
# Time Complexity:
# O(n*n)    ---> Worst Case
# O(nlogn)  ---> Average Case

# Haore's Partition
def Partition(arr,low,high):
    pivot=low
    i=low
    j=low+1
    while j<=high:
        if arr[j]<=arr[pivot]:
            i+=1
            arr[i],arr[j]=arr[j],arr[i]
        j+=1
    arr[i],arr[pivot]=arr[pivot],arr[i]
    return i


def QuickSort(arr,low,high):
    if low<high:
        m=Partition(arr,low,high)
        QuickSort(arr,low,m-1)
        QuickSort(arr,m+1,high)


def main():
    n=int(input())
    arr=list(map(int,input().split()))
    QuickSort(arr,0,n-1)
    print(arr)



main()

File: Days.30/C++/3.Quick-Sort-Array.cpp
// Quick Sort using Haore's Partition

// Time Complexity:
// O(n*n)   ----> Worst Case
// O(nlogn) ----> Best Case and Average Case

#include<iostream>
#include<vector>
using namespace std;


int Partition(vector<int> &arr,int low,int high)
{
    int pivot=low;
    int i=low;
    int j=low+1;
    while (j<=high)
    {
        if (arr[j]<=arr[pivot])
        {
            i+=1;
            swap(arr[i],arr[j]);
        }
        j+=1;
    }
    swap(arr[i],arr[pivot]);
    return i;
}


void QuickSort(vector<int> &arr,int low,int high)
{
    if (low<high)
    {
        int m=Partition(arr,low,high);
        QuickSort(arr,low,m-1);
        QuickSort(arr,m+1,high);
    }
}


int main()
{
    int n;
    cin>>n;
    vector<int> arr(n);
    for (int i=0; i<n; i++)
    {
        cin>>arr[i];
    }
    QuickSort(arr,0,n-1);
    for(int i=0; i<n; i++)
    {
        cout<<arr[i]<<" ";
    }
    cout<<"\n";
    return 0;
}


